src/lxc/caps.c

```diff
@@ -126,6 +147,113 @@ int lxc_caps_up(void)
 	return 0;
 }
 
+int lxc_ambient_caps_up(void)
+{
+	int ret;
+	cap_t caps;
+	cap_value_t cap;
+	char *cap_names = NULL;
+
+	/* When we are run as root, we don't want to play with the capabilities. */
+	if (!getuid())
+		return 0;
+
+	caps = cap_get_proc();
+	if (!caps) {
+		SYSERROR("Failed to retrieve capabilities");
+		return -1;
+	}
+
+	for (cap = 0; cap <= CAP_LAST_CAP; cap++) {
+		cap_flag_value_t flag;
+
+		ret = cap_get_flag(caps, cap, CAP_PERMITTED, &flag);
+		if (ret < 0) {
+			if (errno == EINVAL) {
+				INFO("Last supported cap was %d", cap - 1);
```

Comment:
Don't you want to store the fact that you found the last supported cap here?

Otherwise if lxc was compiled with a CAP_LAST_CAP > the kernel was, then you'll get EPERM below when you try to raise the ambient bit.

Reply:
Ah, excellent, missed that. Thanks, Serge!